Objects/unicodeobject.c

```diff
@@ -15709,17 +15709,37 @@ unicodeiter_next(unicodeiterobject *it)
         int kind = PyUnicode_KIND(seq);
         const void *data = PyUnicode_DATA(seq);
         Py_UCS4 chr = PyUnicode_READ(kind, data, it->it_index);
-        item = PyUnicode_FromOrdinal(chr);
-        if (item != NULL)
-            ++it->it_index;
-        return item;
+        it->it_index++;
+        return unicode_char(chr);
     }
 
     it->it_seq = NULL;
     Py_DECREF(seq);
     return NULL;
 }
 
+static PyObject *
+unicode_ascii_iter_next(unicodeiterobject *it)
+{
+    assert(it != NULL);
+    PyObject *seq = it->it_seq;
+    if (seq == NULL) {
+        return NULL;
+    }
```

Comment:
Is there a code path that can trigger this branch? If not, can we make it an `assert`?

Reply:
I think this should do it:
```python
>>> s = iter('abc')
>>> list(s)
['a', 'b', 'c']
>>> list(s) # here
[]
```

Reply:
In that case, we need a test for that; this branch is not exercised in the test suite.

Reply:
done